### PR TITLE
CUDA mining support

### PR DESCRIPTION
**Pull request by Genoil**
I know you guys are busy reorganizing the whole repo, but after you've completed that work, it shouldn't be too difficult to merge this in.

Some work is still needed on copying the right DLLs into the release folder etc. Also untested on Linux.

### CODE REVIEW

File: ethminer/MinerAux.h [hunk 1]
```diff
@@ -219,6 +245,18 @@ class MinerCLI
...
+		else if (arg == "-U" || arg == "--cuda") {
+			m_minerType = MinerType::CUDA;
+
+			cout << "Genoil's CUDA ethminer" << endl;
+			cout << "=====================================================================" << endl;
+			cout << "Forked from github.com/ethereum/cpp-ethereum" << endl;
+			cout << "Ported from tim hughes' OpenCL kernel" << endl;
+			cout << "With contributions from tpruvot and sp_ " << endl << endl;
+			cout << "Please consider donating a tiny fraction of the extra performance to:" << endl;
+			cout << "ETH: 0xb9310b185455f863f526dab3d245809f6854b4d" << endl;
```
Comment (gavofyork) [inline ethminer/MinerAux.h]: IBAN-35 address?

File: ethminer/MinerAux.h [hunk 2]
```diff
@@ -326,6 +369,31 @@ class MinerCLI
...
+				m_currentBlock
+				))
+				exit(1);
+
+			if (m_cudaDeviceCount != 0) {
+				EthashCUDAMiner::setDevices(m_cudaDevices, m_cudaDeviceCount);
+				m_miningThreads = m_cudaDeviceCount;
+			}
+			EthashCUDAMiner::setNumInstances(m_miningThreads);
+
```
Comment (gavofyork) [inline ethminer/MinerAux.h]: unneeded newline

File: ethminer/MinerAux.h [hunk 3]
```diff
@@ -326,6 +369,31 @@ class MinerCLI
...
+				m_globalWorkSizeMultiplier,
+				m_numStreams,
+				m_openclDevice,
+				m_extraGPUMemory,
+				m_cudaHighCPULoad,
+				m_currentBlock
+				))
+				exit(1);
+
+			if (m_cudaDeviceCount != 0) {
```
Comment (gavofyork) [inline ethminer/MinerAux.h]: bad brace style

File: ethminer/MinerAux.h [hunk 4]
```diff
@@ -219,6 +245,18 @@ class MinerCLI
 			m_minerType = MinerType::CPU;
 		else if (arg == "-G" || arg == "--opencl")
 			m_minerType = MinerType::GPU;
+		else if (arg == "-U" || arg == "--cuda") {
```
Comment (gavofyork) [inline ethminer/MinerAux.h]: bad brace style

File: ethminer/MinerAux.h [hunk 5]
```diff
@@ -219,6 +245,18 @@ class MinerCLI
...
+			m_minerType = MinerType::CUDA;
+
+			cout << "Genoil's CUDA ethminer" << endl;
+			cout << "=====================================================================" << endl;
+			cout << "Forked from github.com/ethereum/cpp-ethereum" << endl;
+			cout << "Ported from tim hughes' OpenCL kernel" << endl;
```
Comment (gavofyork) [inline ethminer/MinerAux.h]: capitalisation for Tim Hughes

File: ethminer/MinerAux.h [hunk 6]
```diff
@@ -168,13 +178,29 @@ class MinerCLI
...
+				try {
+					m_cudaDevices[m_cudaDeviceCount++] = stol(argv[++i]);
+				}
+				catch (...)
+				{
+					break;
+				}
+			}
+		}
+		else if (arg == "--cuda-turbo") {
```
Comment (gavofyork) [inline ethminer/MinerAux.h]: bad brace style

File: ethminer/MinerAux.h [hunk 7]
```diff
@@ -168,13 +178,29 @@ class MinerCLI
...
+			m_cudaHighCPULoad = true;
+		}
+		else if (arg == "--cuda-streams" && i + 1 < argc) {
```
Comment (gavofyork) [inline ethminer/MinerAux.h]: bad brace style

File: ethminer/MinerAux.h [hunk 8]
```diff
@@ -168,13 +178,29 @@ class MinerCLI
...
-#endif
-		else if (arg == "--list-devices")
-			m_shouldListDevices = true;
 		else if (arg == "--allow-opencl-cpu")
 			m_clAllowCPU = true;
-		else if (arg == "--cl-extragpu-mem" && i + 1 < argc)
-			m_extraGPUMemory = 1000000 * stol(argv[++i]);
+#endif
+#if ETH_ETHASHCU || !ETH_TRUE
+		else if (arg == "--cuda-devices") {
```
Comment (gavofyork) [inline ethminer/MinerAux.h]: bad brace style

File: ethminer/MinerAux.h [hunk 9]
```diff
@@ -371,13 +440,22 @@ class MinerCLI
...
 			<< "    --cl-ms-per-batch Set the OpenCL target milliseconds per batch (global workgroup size). Default is " << toString(ethash_cl_miner::c_defaultMSPerBatch) << ". If 0 is given then no autoadjustment of global work size will happen" << endl
 #endif
+#if ETH_ETHASHCU || !ETH_TRUE
+			<< "    --cuda-extragpu-mem Set the memory (in MB) you believe your GPU requires for stuff other than mining. Windows rendering e.t.c.." << endl
+			<< "    --cuda-block-size Set the CUDA block work size. Default is " << toString(ethash_cu_miner::c_defaultBlockSize) << endl
+			<< "    --cuda-grid-size Set the CUDA grid size. Default is " << toString(ethash_cu_miner::c_defaultGridSize) << endl
+			<< "    --cuda-streams Set the number of CUDA streams. Default is " << toString(ethash_cu_miner::c_defaultNumStreams) << endl
+			<< "    --cuda-turbo Get a bit of extra hashrate at the cost of high CPU load... Default is false" << endl
+			<< "    --cuda-devices <0 1 ..n> Select which GPU's to mine on. Default is to use all" << endl
```
Comment (gavofyork) [inline ethminer/MinerAux.h]: grocer's apostrophe.

File: ethminer/MinerAux.h [hunk 10]
```diff
@@ -371,13 +440,22 @@ class MinerCLI
...
+#endif
 			;
 	}
 
 	enum class MinerType
 	{
 		CPU,
-		GPU
+		GPU,
```
Comment (gavofyork) [inline ethminer/MinerAux.h]: rename to `OpenCL` now - `GPU` is misleading.

File: ethminer/MinerAux.h [hunk 11]
```diff
@@ -404,10 +482,13 @@ class MinerCLI
...
 		sealers["opencl"] = GenericFarm<EthashProofOfWork>::SealerDescriptor{&EthashGPUMiner::instances, [](GenericMiner<EthashProofOfWork>::ConstructionInfo ci){ return new EthashGPUMiner(ci); }};
 #endif
+#if ETH_ETHASHCU
+		sealers["cuda"] = GenericFarm<EthashProofOfWork>::SealerDescriptor{ &EthashGPUMiner::instances, [](GenericMiner<EthashProofOfWork>::ConstructionInfo ci){ return new EthashCUDAMiner(ci); } };
+#endif
 		f.setSealers(sealers);
 		f.onSolutionFound([&](EthashProofOfWork::Solution) { return false; });
 
-		string platformInfo = _m == MinerType::CPU ? "CPU" : "GPU";//EthashProofOfWork::CPUMiner::platformInfo() : _m == MinerType::GPU ? EthashProofOfWork::GPUMiner::platformInfo() : "";
+		string platformInfo = _m == MinerType::CPU ? "CPU" : (_m == MinerType::GPU ? "GPU" : "CUDA");
```
Comment (gavofyork) [inline ethminer/MinerAux.h]: `"GPU"` -> `"OpenCL"`

File: ethminer/MinerAux.h [hunk 12]
```diff
@@ -419,6 +500,8 @@ class MinerCLI
 			f.start("cpu");
 		else if (_m == MinerType::GPU)
```
Comment (gavofyork) [inline ethminer/MinerAux.h]: `GPU` -> `CL`

File: ethminer/MinerAux.h [hunk 13]
```diff
@@ -477,6 +560,9 @@ class MinerCLI
 #if ETH_ETHASHCL
 		sealers["opencl"] = GenericFarm<EthashProofOfWork>::SealerDescriptor{&EthashGPUMiner::instances, [](GenericMiner<EthashProofOfWork>::ConstructionInfo ci){ return new EthashGPUMiner(ci); }};
```
Comment (gavofyork) [inline ethminer/MinerAux.h]: `GPU` -> `CL`

File: ethminer/MinerAux.h [hunk 14]
```diff
@@ -491,7 +577,8 @@ class MinerCLI
 			f.start("cpu");
 		else if (_m == MinerType::GPU)
```
Comment (gavofyork) [inline ethminer/MinerAux.h]: `GPU` -> `CL`

File: ethminer/MinerAux.h [hunk 15]
```diff
@@ -586,10 +673,20 @@ class MinerCLI
...
+#endif
 	unsigned m_msPerBatch = ethash_cl_miner::c_defaultMSPerBatch;
 #endif
+#if ETH_ETHASHCU || !ETH_TRUE
+	unsigned m_globalWorkSizeMultiplier = ethash_cu_miner::c_defaultGridSize;
+	unsigned m_localWorkSize = ethash_cu_miner::c_defaultBlockSize;
+	unsigned m_cudaDeviceCount = 0;
+	unsigned m_cudaDevices[16];
+	unsigned m_numStreams = ethash_cu_miner::c_defaultNumStreams;
+	bool	 m_cudaHighCPULoad = false;
```
Comment (gavofyork) [inline ethminer/MinerAux.h]: no tabs/extra spaces within code.

**Comment by gavofyork** [comment]
style work needed. some nomenclature fixes. maybe use qtcreator to cleanse the style of the .h/.cpp files; otherwise check the style guide: brace style particularly incompatible & inconsistent. note always:

```
if (true)
	oneStatement();
else
{
	two();
	statements();
}
```

File: ethminer/MinerAux.h (hunk 1 above)
Comment (Genoil) [inline ethminer/MinerAux.h]: oh crap i'm so bad at git. the txt isn't supposed to in the PR...but what did you mean anyway?

File: ethminer/MinerAux.h [hunk 16]
```diff
@@ -404,10 +485,13 @@ class MinerCLI
 #if ETH_ETHASHCL
 		sealers["opencl"] = GenericFarm<EthashProofOfWork>::SealerDescriptor{&EthashGPUMiner::instances, [](GenericMiner<EthashProofOfWork>::ConstructionInfo ci){ return new EthashGPUMiner(ci); }};
 #endif
+#if ETH_ETHASHCU
+		sealers["cuda"] = GenericFarm<EthashProofOfWork>::SealerDescriptor{ &EthashGPUMiner::instances, [](GenericMiner<EthashProofOfWork>::ConstructionInfo ci){ return new EthashCUDAMiner(ci); } };
```
Comment (gavofyork) [inline ethminer/MinerAux.h]: should probably be `&EthashCUDAMiner::instances`.

File: ethminer/MinerAux.h [hunk 17]
```diff
@@ -404,10 +485,13 @@ class MinerCLI
...
+#endif
 		f.setSealers(sealers);
 		f.onSolutionFound([&](EthashProofOfWork::Solution) { return false; });
 
-		string platformInfo = _m == MinerType::CPU ? "CPU" : "GPU";//EthashProofOfWork::CPUMiner::platformInfo() : _m == MinerType::GPU ? EthashProofOfWork::GPUMiner::platformInfo() : "";
+		string platformInfo = _m == MinerType::CPU ? "CPU" : (_m == MinerType::CL ? "GPU" : "CUDA");
```
Comment (gavofyork) [inline ethminer/MinerAux.h]: `"GPU"` -> `"CL"`

File: libethcore/EthashCUDAMiner.cpp [hunk 18]
```diff
@@ -0,0 +1,237 @@
...
+}
+
+unsigned EthashCUDAMiner::s_platformId = 0;
+unsigned EthashCUDAMiner::s_deviceId = 0;
+unsigned EthashCUDAMiner::s_numInstances = 0;
+int EthashCUDAMiner::s_devices[16] = { -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1 };
+
+
+EthashCUDAMiner::EthashCUDAMiner(ConstructionInfo const& _ci) :
+GenericMiner<EthashProofOfWork>(_ci),
```
Comment (gavofyork) [inline libethcore/EthashCUDAMiner.cpp]: tabbing.

File: libethcore/EthashCUDAMiner.cpp [hunk 19]
```diff
@@ -0,0 +1,237 @@
...
+
+	}
+}
+
+unsigned EthashCUDAMiner::s_platformId = 0;
+unsigned EthashCUDAMiner::s_deviceId = 0;
+unsigned EthashCUDAMiner::s_numInstances = 0;
+int EthashCUDAMiner::s_devices[16] = { -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1 };
+
+
```
Comment (gavofyork) [inline libethcore/EthashCUDAMiner.cpp]: unneeded newline.

File: libethcore/EthashCUDAMiner.cpp [hunk 20]
```diff
@@ -0,0 +1,237 @@
...
+	return ethash_cu_miner::listDevices();
+}
+
+bool EthashCUDAMiner::configureGPU(
+	unsigned _blockSize,
+	unsigned _gridSize,
+	unsigned _numStreams,
+	unsigned _deviceId,
+	unsigned _extraGPUMemory,
+	bool	 _highcpu,
```
Comment (gavofyork) [inline libethcore/EthashCUDAMiner.cpp]: no tabs or extra spaces within code.

File: libethcore/EthashCUDAMiner.h [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+#pragma once
+#if ETH_ETHASHCU || !ETH_TRUE
+
+#include "libdevcore/Worker.h"
+#include "EthashAux.h"
+#include "Miner.h"
+
+namespace dev
+{
+	namespace eth
```
Comment (gavofyork) [inline libethcore/EthashCUDAMiner.h]: no namespace indentation.

File: libethcore/EthashCUDAMiner.cpp [hunk 22]
```diff
@@ -0,0 +1,237 @@
...
+#include <thread>
+#include <chrono>
+#include <libethash-cu/ethash_cu_miner.h>
+using namespace std;
+using namespace dev;
+using namespace eth;
+
+namespace dev
+{
+	namespace eth
```
Comment (gavofyork) [inline libethcore/EthashCUDAMiner.cpp]: no namespace indentation

File: libethcore/EthashCUDAMiner.h [hunk 23]
```diff
@@ -0,0 +1,86 @@
...
+			static std::string platformInfo();
+			static unsigned getNumDevices();
+			static void listDevices();
+			static bool configureGPU(
+				unsigned _blockSize,
+				unsigned _gridSize,
+				unsigned _numStreams,
+				unsigned _deviceId,
+				unsigned _extraGPUMemory,
+				bool	 _highcpu,
```
Comment (gavofyork) [inline libethcore/EthashCUDAMiner.h]: no tabs/spaces within code.

File: libethcore/EthashCUDAMiner.h [hunk 24]
```diff
@@ -0,0 +1,86 @@
...
+				uint64_t _currentBlock
+				);
+			static void setNumInstances(unsigned _instances) { s_numInstances = std::min<unsigned>(_instances, getNumDevices()); }
+			static void setDevices(unsigned * _devices, unsigned _selectedDeviceCount) {
```
Comment (gavofyork) [inline libethcore/EthashCUDAMiner.h]: bad braces.

File: libethcore/EthashCUDAMiner.h [hunk 25]
```diff
@@ -0,0 +1,86 @@
...
+				for (unsigned i = 0; i < _selectedDeviceCount; i++) {
```
Comment (gavofyork) [inline libethcore/EthashCUDAMiner.h]: bad braces.

File: libethcore/EthashCUDAMiner.h [hunk 26]
```diff
@@ -0,0 +1,86 @@
...
+			using GenericMiner<EthashProofOfWork>::accumulateHashes;
+
+			EthashCUHook* m_hook = nullptr;
+			ethash_cu_miner* m_miner = nullptr;
+
+			h256 m_minerSeed;		///< Last seed in m_miner
+			static unsigned s_platformId;
+			static unsigned s_deviceId;
+			static unsigned s_numInstances;
+			static int		s_devices[16];
```
Comment (gavofyork) [inline libethcore/EthashCUDAMiner.h]: no tabs/spaces within code.

File: libethcore/EthashSealEngine.cpp [hunk 27]
```diff
@@ -35,6 +36,9 @@ EthashSealEngine::EthashSealEngine()
 #if ETH_ETHASHCL
 	sealers["opencl"] = GenericFarm<EthashProofOfWork>::SealerDescriptor{&EthashGPUMiner::instances, [](GenericMiner<EthashProofOfWork>::ConstructionInfo ci){ return new EthashGPUMiner(ci); }};
 #endif
+#if ETH_ETHASHCU
```
Comment (gavofyork) [inline libethcore/EthashSealEngine.cpp]: would really prefer `ETH_ETHASHCUDA`

**Comment by Genoil** [comment]
Not sure if you guys really need this CUDA miner after all. Applied some tweaks to the OpenCL kernel, making it as fast as the CUDA kernel...

https://github.com/Genoil/cpp-ethereum/tree/cuda-opencl-ptx

**Comment by gavofyork** [comment]
Maybe make a PR for the tweaked kernel (esp if it doesn't impact AMD GPUs)?

**Comment by Genoil** [comment]
I've sent it to Tim for him to check. Haven't got AMD hw..btw i'll keep the CUDA miner, it is still few % faster.

**Pull request closed without merging**